### ISSUE TITLE
AssetManager not work correctly with Relative URL in getAssetUrl method by assetMap.

### ISSUE THREAD

**Issue by porshkevich** [yiisoft/yii2#8522]
In base by assetMap we can change asset url to Relative URL.
But AssetManager use only 2 rule for that changing
1. Base rule. (BaseUrl + AssetUrl)
2. A relative URL that begins with // (two slashes) always replaces everything from the hostname onwards. 

and forget about the third rule
3. A relative URL that begins with / (a slash) always replaces the entire pathname of the base URL. 

I propose to change
```php
public function getAssetUrl($bundle, $asset)
    {
        if (($actualAsset = $this->resolveAsset($bundle, $asset)) !== false) {
            return Url::isRelative($actualAsset) ? $this->baseUrl . '/' . $actualAsset : $actualAsset;
        } else {
            return Url::isRelative($asset) ? $bundle->baseUrl . '/' . $asset : $asset;
        }
    }
```
to
```php
public function getAssetUrl($bundle, $asset)
    {
        if (($actualAsset = $this->resolveAsset($bundle, $asset)) !== false) {
            return Url::isRelative($actualAsset) && strncmp($actualAsset, '/', 1) ? $this->baseUrl . '/' . $actualAsset : $actualAsset;
        } else {
            return Url::isRelative($asset) ? $bundle->baseUrl . '/' . $asset : $asset;
        }
    }
```
This will be used for such substitution by assetMap like

assetMap value | genaretad url | result url
-------------------- | ----------------- | -----------
js/jquery.js | $baseUrl/js/jquery.js | http://sitehost/path/to/baseUrl/js/jquery.js
/js/jquery.js | /js/jquery.js | http://sitehost/js/jquery.js
//ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js | //ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js | http://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js

**Comment by klimov-paul** [comment]
I am unsure.
Although such change can be made it will encourage bad practice. Using domain relative paths (starting from '/') will produce invalid results in case project is located not under domain root, e.g. 'http://domain.com/myproject'.

Also feature actually breaks a BC. If asset is placed in `assetMap` with leading '/' it produces URL with double '/', e.g. 'http://domain.com/assets//css/overrride.css', which is resolved by browser correctly as 'http://domain.com/assets/css/overrride.css'. If someone is making mistake placing leading '/' in asset of `assetMap` he may not realize this.

Current walkaround is using folder relative URL with '..'. For example, asset specified as '../js/jquery_overridde.js' will produce URL '/assets/../js/jquery_overridde.js', which will be resolved as 'http://domain.com/js/jquery_overridde.js'.

**Comment by porshkevich** [comment]
> in case project is located not under domain root, e.g. 'http://domain.com/myproject'

in this case, those who will use "/" will be able to avoid this error, because they already know they need URL (or you can use even so site`s baseUrl, instead asset`s baseUrl)

> If someone is making mistake placing leading '/' in asset of assetMap he may not realize this.

double slashes repeatedly discussed on the Internet, and this is not normal behavior of any system or browser. if the user makes a mistake, it should lead to error, and do not pretend that a minor error.

> Current walkaround is using folder relative URL with '..'. For example, asset specified as '../js
/jquery_overridde.js' will produce URL '/assets/../js/jquery_overridde.js', which will be resolved as 'http://domain.com/js/jquery_overridde.js'.

I agree that my version is not the most elegant and not the fact that the best, but the use of "..", for me it does not look better.

**Comment by klimov-paul** [comment]
> in this case, those who will use "/" will be able to avoid this error, because they already know they need URL

In this case result varies by environment. Following your logic you may just specify absolute URL for your 'jquery.js' just as well.
Relying on domain relative path (starting from '/') is not a good practice. And it does note metter where it is used: for `assetMap` or background-url specifications inside CSS files.
Someone may as well expect framework will automatically prepend such URL with `UrlManager::baseUrl`.

> double slashes repeatedly discussed on the Internet, and this is not normal behavior of any system or browser. if the user makes a mistake, it should lead to error, and do not pretend that a minor error.

I have never told it is a valid usage. But if the fix is going to be made it should be reflected at `UPGRADE.md` just in case. Or postponed to 2.1.

> I agree that my version is not the most elegant and not the fact that the best, but the use of "..", for me it does not look better.

Still, until solution is not provided you may use it. At least it does not depend on the environment and will work the same no mattter if project located under sub directory or not.
Or you can specify absolute URL for your asset.

**Comment by porshkevich** [comment]
> Following your logic you may just specify absolute URL for your 'jquery.js' just as well.

Regarding me, I am already using the proposed solution. Because I can`t be an absolute Url (multiple domains). And forgive but not for 'jquery.js', it was listed as an example.

I just decided to bring the idea up for discussion, because I believe it can be useful to others.

**Comment by porshkevich** [comment]
> Someone may as well expect framework will automatically prepend such URL with UrlManager::baseUrl

If need this, can use another code (already adapted to the latest release):
```php
public function getAssetUrl($bundle, $asset)
    {
        if (($actualAsset = $this->resolveAsset($bundle, $asset)) !== false) {
            $asset = $actualAsset;
			if (strncmp($asset, '/', 1)) {
				$basePath = $this->basePath;
				$baseUrl = $this->baseUrl;
			}
			else {
				$asset = ltrim($asset, '/');
				$basePath = "@webroot";
				$baseUrl = "@web";
			}
        } else {
            $basePath = $bundle->basePath;
            $baseUrl = $bundle->baseUrl;
        }

        if (!Url::isRelative($asset)) {
            return $asset;
        }

        if ($this->appendTimestamp && ($timestamp = @filemtime("$basePath/$asset")) > 0) {
            return "$baseUrl/$asset?v=$timestamp";
        } else {
            return "$baseUrl/$asset";
        }
    }
```

**Comment by klimov-paul** [comment]
Issue resolved by commit 95ba06c
